### PR TITLE
Multistage container image based on UBI

### PR DESCRIPTION
**Pull request by enkeys**

### CODE REVIEW

**Review by jiridanek — APPROVED** [review]
looks good, just some discussion points as line comments

also, note: we could delete compile_java.sh because that is now unused; or do something else with it, but i'm ok leaving it for later

File: Dockerfile [hunk 1]
```diff
@@ -1,27 +1,53 @@
-FROM fedora:latest
+# Arguments for DEV's (comment static FROM and uncomnnet #DEV ones)
+ARG UBI_VERSION=9
+ARG OPENJDK_VERSION=17
+ARG UBI_BUILD_TAG=latest
+ARG UBI_RUNTIME_TAG=latest
+ARG IMAGE_BUILD=registry.access.redhat.com/ubi${UBI_VERSION}/openjdk-${OPENJDK_VERSION}:${UBI_TAG}
```
Comment (jiridanek) [inline Dockerfile:6]: does this run the image build repeatedly for each arch, incl downloading maven artifacts? is it possible to improve on this in future? either cache artifacts, or do one build for all the multiarch images?

File: Dockerfile [hunk 2]
```diff
@@ -1,27 +1,53 @@
-FROM fedora:latest
+# Arguments for DEV's (comment static FROM and uncomnnet #DEV ones)
```
Comment (jiridanek) [inline Dockerfile:1]: aha, got it, disregard comment on another #DEV line

I did not notice this line before

Comment (enkeys) [reply Dockerfile:1]: I was thinking about two container files for images, finally I decided to just insert this comment.  I'm not sure if we want to leave it there or not, but it's useful from time to time.😄

File: Dockerfile (hunk 1 above)
Comment (enkeys) [reply Dockerfile:6]: Yes, there is possibility to use GitHub Action Cache for Maven and use it in every job. Then in the image we can reuse it in builder stage and take cache from the path  `COPY . /app`

File: Dockerfile [hunk 3]
```diff
@@ -1,27 +1,53 @@
...
-RUN yum install \
-    -y --setopt=install_weak_deps=0 --setopt=tsflags=nodocs \
-    java-11-openjdk-headless \
-    bzip2 unzip xz \
-    bzr git mercurial openssh-clients subversion procps \
-    gnupg dirmngr \
-    ca-certificates curl wget \
-    && dnf clean all -y
+#DEV FROM $IMAGE_BASE
+FROM registry.access.redhat.com/ubi9/openjdk-17-runtime:1.13-8
```
[the remark this answers is not in the document]
Comment (enkeys) [reply Dockerfile:28]: It's required for consistent build and for security. When you are aware of CVE's in used version, you have chance to catch up and do upgrade. And Dependabot, if we have enough tests, what about automatic merge? 😃 And I'm personally happier with more commits when it comes to upgrading dependency.

[the remark this answers is not in the document]
Comment (jiridanek) [reply Dockerfile:28]: We don't have any tests that go through the container...

Ok, let's go with precise versions, although I doubt that it does really matter for our usage. CVEs is a fraud!

* https://www.theregister.com/2021/07/09/npm_audit_security/
* https://pkg.go.dev/golang.org/x/vuln/cmd/govulncheck

File: Dockerfile (hunk 1 above)
Comment (jiridanek) [reply Dockerfile:6]: 💌

File: create_links.sh [hunk 4]
```diff
@@ -1,34 +0,0 @@
...
-for j in sender receiver connector; do
-    filename="/usr/local/bin/${i%.jar}-${j}"
-    cat > "${filename}" << EOF
-#!/bin/sh
-
-java \${JAVA_OPTS} \${CLI_QPID_JMS_OPTS} -jar /main/${i} ${j} \$@
-EOF
-    chmod +x "$filename"
-  done
-done
```
[the remark this answers is not in the document]
Comment (enkeys) [reply create_links.sh:34]: Yup, understand, I removed it for simpler build without looking to a script. But we can definitely discuss it.

I see one difference right now, look at: 
- [ ] ` \${CLI_QPID_JMS_OPTS}` 

So should we use for every client different env, or one $CLI_JAVA_OPTS for example?

[the remark this answers is not in the document]
Comment (jiridanek) [reply create_links.sh:34]: Maybe let it use the usual `JAVA_OPTS` that every JVM looks at by default? ... There might be issue of backwards compatibility, some ssl tests using the images were probably setting the existing variables; if we rename it, we might have to propagate the changes

File: Dockerfile [hunk 5]
```diff
@@ -1,27 +1,53 @@
...
 
-WORKDIR /var/lib/cli-java
+#DEV FROM $IMAGE_BUILD AS build
+FROM registry.access.redhat.com/ubi9/openjdk-17:1.13-9 AS build
+
+USER root
+COPY . /app
+WORKDIR /app
+
+ENV MAVEN_OPTS="-XX:+TieredCompilation -XX:TieredStopAtLevel=1 -Dmaven.repo.local=/app/.m2 -Dmaven.artifact.threads=42"
```
[the remark this answers is not in the document]
Comment (enkeys) [reply Dockerfile:16]: We need benchmark! 🍭  Yeah, I must agree, but it was nice to learn something about java.  😄 So do we want to take it away?

**Comment by jiridanek** [comment]
@enkeys looks fine to me, feel free to merge and post the gha PR ;P

**Comment by jiridanek** [comment]
> @enkeys looks fine to me, feel free to merge and post the gha PR ;P

@enkeys ping? is this ready?

**Comment by jiridanek** [comment]
Shipshape issue https://github.com/rh-messaging/shipshape/issues/63

**Comment by enkeys** [comment]
@jiridanek  pong " I had almost forgotten about this PR" 🥸  and now everyone's pressuring me, so let's merge it before disaster strikes. 🚀

**Comment by enkeys** [comment]
Fixes #484 and closes PR #485.

**Pull request merged**